src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -0,0 +1,210 @@
+                declaration.Modifiers.Any(m => m.IsKind(SyntaxKind.PartialKeyword)) &&
+                !GeneratorHelper.IsWinRTType(declaration); // Making sure it isn't an RCW we are projecting.
+        }
+
+        private static VtableAttribute GetVtableAttributeToAdd(GeneratorSyntaxContext context)
+        {
+            var symbol = context.SemanticModel.GetDeclaredSymbol(context.Node as ClassDeclarationSyntax);
+
+            HashSet<string> interfacesToAddToVtable = new();
+            foreach (var iface in symbol.Interfaces)
+            {
+                AddInterfaceAndCompatibleInterfacesToVtable(iface);
+
+                foreach (var baseIface in iface.AllInterfaces)
+                {
+                    AddInterfaceAndCompatibleInterfacesToVtable(baseIface);
+                }
+            }
+
+            bool hasWinRTExposedBaseType = false;
+            var baseType = symbol.BaseType;
+            while (baseType != null)
+            {
+                if (GeneratorHelper.IsWinRTType(baseType) ||
+                    baseType.Interfaces.Any(GeneratorHelper.IsWinRTType))
+                {
+                    hasWinRTExposedBaseType = true;
+                    break;
+                }
+
+                baseType = baseType.BaseType;
+            }
+
+            return new VtableAttribute(
+                symbol.ContainingNamespace.ToDisplayString(),
+                symbol.ContainingNamespace.IsGlobalNamespace,
+                symbol.Name,
+                interfacesToAddToVtable.ToArray(),
+                hasWinRTExposedBaseType);
+
+            void AddInterfaceAndCompatibleInterfacesToVtable(INamedTypeSymbol iface)
+            {
+                if (GeneratorHelper.IsWinRTType(iface))
+                {
+                    interfacesToAddToVtable.Add(iface.ToDisplayString());
+                }
+
+                if (iface.IsGenericType && TryGetCompatibleWindowsRuntimeTypesForVariantType(iface, null, out var compatibleIfaces))
+                {
+                    foreach (var compatibleIface in compatibleIfaces)
+                    {
+                        interfacesToAddToVtable.Add(compatibleIface.ToDisplayString());
+                    }
+                }
+            }
+        }
+
+        private static bool TryGetCompatibleWindowsRuntimeTypesForVariantType(INamedTypeSymbol type, Stack<INamedTypeSymbol> typeStack, out IList<INamedTypeSymbol> compatibleTypes)
+        {
+            compatibleTypes = null;
+
+            // Out of all the C# interfaces which are valid WinRT interfaces and
+            // support covariance, they all only have one generic parameter,
+            // so scoping to only handle that.
+            if (!(type.IsGenericType &&
+                type.TypeParameters.Length == 1 &&
+                type.TypeParameters[0].Variance == VarianceKind.Out &&
+                !type.TypeArguments[0].IsValueType))

```

Comment:
Nit: consider using a pattern to simplify things?
```suggestion
            if (type is not { IsGenericType: true, TypeParameters: [{ Variance: VarianceKind.Out, IsValueType: false })
```
You can add PolySharp as a package reference to this project to get all the necessary polyfills, then you can just set the language version to C# 11 in the generator and it'll work fine 🙂